### PR TITLE
arm_update_activate.sh: Update to new payload format

### PR DESCRIPTION
**Pull request by jonathanhaigh-arm**
We're going to be using swupdate for part of our firmware update flow,
so change our payload format to something that swupdate understands.
I.e. a cpio archive containing a file for each component.

I've also changed the formats of some of the individual component
payloads:
* Payloads for bootloader components are no longer .tar.xz files. They
  just contain a single file so there's no need for the tar archive and
  they are now just plain .xz files.
* The payload for the rootfs component is now called ROOTFS.tar.xz to
  match the naming scheme of other components.
* Instead of having individual .ipk files in the overall payload
  archive, there's now an APPS.tar.xz file that contains all of the
  apps.

Our next step in integrating swupdate will be to pass swupdate the
overall payload archive and let it delegate to scripts for each of the
components. Each component will only recieve its own component payload,
not the overall payload archive, so I've refactored
arm_update_activate.sh to work a bit more like that: now, we have a
function for each component type and that function gets passed just the
individual component payload.

I've factored the extract_and_write_update_component_or_die() function
into smaller ones because I found it confusing, and, now that bootloader
components aren't in tar files, it makes less sense to have a single
function that deals with both boot partition updates and bootloader
updates.

I've factored out a couple of functions for getting sizes in bytes from
files that contain sizes in KiB.

I started finding it hard to track whether the script was cleaning up
after itself so:
* Create a new temporary directory that the script uses so we can clean
  up by just removing that directory.
* Run a cleanup function that removes the temporary directory whenever
  the script exits.

I've also removed some early exits that happen after an individual
component has been successfully installed. This allows multiple
components to be updated using a single call to the script.

For IOTMBL-2540: Robust update: switch to using swupdate's payload
format

**Testing, Lava links and Jenkins links**
* See https://github.com/ARMmbed/meta-mbl/pull/730.

**Related PRs**
* meta-mbl PR to update create-update-payload: https://github.com/ARMmbed/meta-mbl/pull/730.
* mbl-tools PR to update build-update-payloads.py: ARMmbed/mbl-tools#334.

### CODE REVIEW

File: cloud-services/mbl-cloud-client/scripts/arm_update_activate.sh [hunk 1]
```diff
@@ -193,171 +155,229 @@ gdnflod_label="$1"
...
+# $4: Set to "yes" if non-existence of the partition is not a failure
+update_boot_part_from_dir_or_die() {
+ubpd_part_name="$1"
+ubpd_dir="$2"
+ubpd_size_estimate_B="$3"
+ubpd_skip_ok="$4"
+
+    if is_part_skipped "$ubpd_part_name"; then
+        if [ "$ubpd_skip_ok" != "yes" ]; then
+            printf "Partition \"%s\" does not exist\n" "%ubpd_part_name"
```
Comment (rwalton-arm) [inline cloud-services/mbl-cloud-client/scripts/arm_update_activate.sh]:
```suggestion
            printf "Partition \"%s\" does not exist\n" "$ubpd_part_name"
```

File: cloud-services/mbl-cloud-client/scripts/arm_update_activate.sh [hunk 2]
```diff
@@ -488,13 +456,127 @@ elif ROOTFS_FILE=$(echo "${FIRMWARE_FILES}" | grep '^rootfs\.tar\.xz$'); then
...
+
+extract_all_files_from_payload_or_die "$FIRMWARE_PATH" "$PAYLOAD_TMP_DIR"
+ensure_payload_format_support_or_die "${PAYLOAD_TMP_DIR}/payload_format_version"
+
+# We only have to reboot during the update if there's an update for a component
+# that requires a reboot, so set the do_not_reboot flag here and let it be
+# deleted if a reboot is required..
+touch "${TMP_DIR}/do_not_reboot"
+
+# list files from udpate payload
```
Comment (rwalton-arm) [inline cloud-services/mbl-cloud-client/scripts/arm_update_activate.sh]:
```suggestion
# list files from update payload
```

File: cloud-services/mbl-cloud-client/scripts/arm_update_activate.sh [hunk 3]
```diff
@@ -20,7 +20,6 @@
 # Note - exit codes 1-20 come from arm_udpate_common.sh.
```
Comment (rwalton-arm) [inline cloud-services/mbl-cloud-client/scripts/arm_update_activate.sh]:
```suggestion
# Note - exit codes 1-20 come from arm_update_common.sh.
```

**Comment by jonathanhaigh-arm** [comment]
I have some provisional changes to udpate the tests that create update payloads on the [jh-cpio-test-IOTMBL-2540](https://github.com/ARMmbed/mbl-core/tree/jh-cpio-test-IOTMBL-2540) branch that I'm testing now, but I'll probably move those changes here because they should be merged at the same time as the changes here in order to not break the tests.

**Comment by jonathanhaigh-arm** [comment]
> I have some provisional changes to udpate the tests that create update payloads on the jh-cpio-test-IOTMBL-2540 branch that I'm testing now, but I'll probably move those changes here because they should be merged at the same time as the changes here in order to not break the tests.

I'm abandoning this work. We're going to use the `create-update-payload` script from meta-mbl to create the updates. We'll have to move the payload generation into Jenkins from Lava because `create-update-payload` requires the BitBake build environment. That should be straightforward for the rootfs payload - we already have the rootfs .tar.xz lying around in the build area. We don't currently have access to the apps from our BitBake build environment, but somebody will work on moving the app compilation from Lava into Jenkins. In the mean time, I'm going to break the app update tests.

**Pull request merged**